src/main/java/com/google/sps/data/FirestoreDB.java

```diff
@@ -104,23 +112,117 @@ public void followTag(String userId, String tagId) {
     return;
   }
 
-  public Iterable<RecipeMetadata> getRecipesMatchingTags(Iterable<String> tagIds) {
-    return null;
+  public List<RecipeMetadata> getRecipesMatchingTags(List<String> tagIds) {
+    Query query = recipesMatchingTags(tagIds, tagIds.iterator());
+    return getRecipeMetadataQuery(query);
   }
-  public Iterable<RecipeMetadata> getRecipesMatchingCreator(String creatorId) {
-    return null;
+
+  public List<RecipeMetadata> getRecipesMatchingCreator(String creatorId) {
+    Query query = DBUtils.recipes().whereEqualTo("creatorId", creatorId);
+    return getRecipeMetadataQuery(query);
   }
-  public Iterable<RecipeMetadata> getRecipesMatchingIDs(Iterable<String> Ids) {
-    return null;
+
+  public List<RecipeMetadata> getRecipesSavedBy(String userId) {
+    List<String> saved_Ids = savedRecipeIds(userId);
+    return getRecipesMatchingIDs(saved_Ids);
   }
 
-  public Comment addComment(Comment newComment) {
-    return null;
+  public List<RecipeMetadata> getRecipesMatchingIDs(List<String> Ids) {
+    ArrayList<String> idList = new ArrayList<>();
+    Iterator<String> iter = Ids.iterator();
+    while (iter.hasNext()) {
```

Comment:
this code block seems really odd to me-- is there any reason we can't just query `DBUtils.recipes().whereIn(Recipe.ID_KEY, idList)` ?

Reply:
The parameter Ids was an Iterable before, this has been changed now